pkg/extensions/sync/utils.go

```diff
@@ -363,18 +355,33 @@ func getImageRef(registryDomain, repo, tag string) (types.ImageReference, error)
 
 // get a local ImageReference used to temporary store one synced image.
 func getLocalImageRef(imageStore storage.ImageStore, repo, tag string) (types.ImageReference, string, error) {
-	uuid, err := guuid.NewV4()
-	// hard to reach test case, injected error, see pkg/test/dev.go
-	if err := test.Error(err); err != nil {
-		return nil, "", err
-	}
+	// check for existing local repo folder before creating a new one
+	localRepoPath := path.Join(imageStore.RootDir(), repo, SyncBlobUploadDir)
+	if fileList, err := os.ReadDir(localRepoPath); os.IsNotExist(err) || len(fileList) == 0 {
+		// no pre-existing local cache for repo, creating folder
+		uuid, err := guuid.NewV4()
+		// hard to reach test case, injected error, see pkg/test/dev.go
+		if err := test.Error(err); err != nil {
+			return nil, "", err
+		}
 
-	localCachePath := path.Join(imageStore.RootDir(), repo, SyncBlobUploadDir, uuid.String())
+		localCachePath := path.Join(imageStore.RootDir(), repo, SyncBlobUploadDir, uuid.String())
 
-	if err = os.MkdirAll(path.Join(localCachePath, repo), storage.DefaultDirPerms); err != nil {
+		if err = os.MkdirAll(path.Join(localCachePath, repo), storage.DefaultDirPerms); err != nil {
+			return nil, "", err
+		}
+	}
+	// local cache for repo
+	fileList, err := os.ReadDir(localRepoPath)
+	if err != nil {
 		return nil, "", err
 	}
+	// using the existing one (which must be a dir and only one)
+	if !fileList[0].IsDir() || len(fileList) > 1 {
+		panic("Something wrong with the existing repo dir: " + localRepoPath)
```

Comment:
this will fail in some scenarios, if a periodically sync is running on image A and a on demand runs on the same image A with another tag it will fail because there will be more than one sync subdir.

there should be a better way to treat this, maybe splitting this function in two?

one for getting a localCachePath given a repo
one for getting a localImageRef given a localCachePath and a tag

Reply:
Split getLocalImageRef function by adding a getLocalCachePath that generates a unique folder so parallel execution will not be affected